server/src/main/java/org/elasticsearch/index/mapper/SourceLoader.java

```diff
@@ -104,7 +106,23 @@ public Set<String> requiredStoredFields() {
         @Override
         public Leaf leaf(LeafReader reader, int[] docIdsInLeaf) throws IOException {
             SyntheticFieldLoader loader = syntheticFieldLoaderLeafSupplier.get();
-            return new SyntheticLeaf(loader, loader.docValuesLoader(reader, docIdsInLeaf));
+            return new LeafWithMetrics(new SyntheticLeaf(loader, loader.docValuesLoader(reader, docIdsInLeaf)));
+        }
+
+        private record LeafWithMetrics(Leaf leaf) implements Leaf {
+
+            @Override
+            public Source source(LeafStoredFieldLoader storedFields, int docId) throws IOException {
+                SourceFieldMetrics metrics = MapperMetrics.SOURCE_FIELD_METRICS.get();
+                long startTime = metrics.getRelativeTimeSupplier().getAsLong();
+
+                var source = leaf.source(storedFields, docId);
+
+                TimeValue duration = TimeValue.timeValueMillis(metrics.getRelativeTimeSupplier().getAsLong() - startTime);
```

Comment:
Since start time is also coming from metrics, consider moving this inside SourceFieldMetrics and changing the recording function to:

`void recordSyntheticSourceLoadLatency(long endTimeMillis)`

Reply:
`SourceFieldMetrics` does not have any state, it only has a `Supplier` of time, not the time itself.

Reply:
Yeah good point, we'd still have to pass the start time so this doesn't help much. Sounds good.